src/c++/libos/lwip/lwip_queue.cc

```diff
@@ -1399,35 +1481,42 @@ dmtr::lwip_queue::parse_packet(struct sockaddr_in &src,
     dst.sin_family = AF_INET;
 
     // segment count
-    sga.sga_numsegs = ntohl(*reinterpret_cast<uint32_t *>(p));
-    p += sizeof(uint32_t);
+    uint32_t numsegs_buf;
+    auto * const numsegs = pktmbuf_struct_read(pkt, offset, numsegs_buf);
+    sga.sga_numsegs = ntohl(*numsegs);
+    offset += sizeof(uint32_t);
+
 
 #if DMTR_DEBUG
     printf("recv: sga_numsegs: %d\n", sga.sga_numsegs);
 #endif
 
     if (sga.sga_numsegs != 0xdeadbeef) {
+        // Allocate enough space for all sga segments
+        size_t sga_total_size = ntohs(ip_hdr->total_length) - sizeof(*ip_hdr) - sizeof(*udp_hdr) - sizeof(uint32_t);
+        DMTR_OK(dmtr_malloc(&sga.sga_buf, sga_total_size));
+
+        // Read into the sga buffer
+        const void *read_buf = rte_pktmbuf_read(pkt, offset, sga_total_size, sga.sga_buf);
+        // If read_buf is not pointing to sga_buf, it's because no copy was necessary
+        // (to avoid freeing problems later, copy into sga_buf anyway)
+        if (read_buf != sga.sga_buf) {
+            rte_memcpy(sga.sga_buf, read_buf, sga_total_size);
+        }
+
+        char *offset_buf = static_cast<char*>(sga.sga_buf);
         for (size_t i = 0; i < sga.sga_numsegs; ++i) {
-            // segment length
-            auto seg_len = ntohl(*reinterpret_cast<uint32_t *>(p));
+            uint32_t seg_len = ntohl(*(uint32_t*)offset_buf);
+            offset_buf += sizeof(seg_len);
```

Comment:
Isn't it clearer to explicitly give `uint32_t` as an argument to `sizeof()`? I am always confused with regard to best practices here.

Reply:
My thoughts are that the pointer should be advanced by the size of `seg_len`, regardless of what type `seg_len` is. If we later change it to be a uint16_t, then the pointer should advance by that much, and we'd have to change code in two places to make that change. This way, we just have to change the type of seg_len directly.
I'm not incredibly tied to it though. I don't think there are strict guidelines on it -- especially in c++, where this type of pointer arithmetic is sorta frowned upon anyway.